File: Go/Pow(x, n).go
// 0ms, 100.00%; 2.2MB, 7.01%
func myPow(x float64, n int) float64 {
	if n < 0 {
		return 1.0 / helper(x, -int64(n))
	}
	return helper(x, int64(n))
}

func helper(x float64, n int64) float64 {
	ret := 1.0
	if (n & 1) == 1 {
		ret = x
		n--
	}
	switch n {
	case 0:
		return ret
	case 1:
		return ret * x
	case 2:
		return ret * x * x
	default:
		return ret * helper(helper(x, n/2), 2)
	}
}

File: Java/Pow(x, n).java
// 1ms, 66.84%; 40.3MB, 5.01%
class Solution {
    public double myPow(double x, int n) {
        if (n < 0) {
            return 1.0 / helper(x, Math.abs((long) n));
        }
        return helper(x, n);
    }

    private double helper(double x, long n) {
        double ret = 1.0;
        if ((n & 1) == 1) {
            ret = x;
            --n;
        }
        return ret * (n == 0 ? 1.0 : n == 1 ? x : n == 2 ? x * x : helper(helper(x, n / 2), 2));
    }
}
